store/tikv/latch/latch.go

```diff
@@ -22,32 +23,26 @@ import (
 	"github.com/spaolacci/murmur3"
 )
 
-// latch stores a key's waiting transactions information.
-type latch struct {
-	// Whether there is any transaction in waitingQueue except head.
-	hasMoreWaiting bool
-	// The startTS of the transaction which is the head of waiting transactions.
-	waitingQueueHead uint64
-	maxCommitTS      uint64
-	sync.Mutex
-}
+type node struct {
+	slotID      int
+	key         []byte
+	maxCommitTS uint64
+	value       *Lock
 
-func (l *latch) isEmpty() bool {
-	return l.waitingQueueHead == 0 && !l.hasMoreWaiting
+	next *node
 }
 
-func (l *latch) free() {
-	l.waitingQueueHead = 0
-}
-
-func (l *latch) refreshCommitTS(commitTS uint64) {
-	l.Lock()
-	defer l.Unlock()
-	l.maxCommitTS = mathutil.MaxUint64(commitTS, l.maxCommitTS)
+// latch stores a key's waiting transactions information.
+type latch struct {
+	queue   *node
+	count   int
+	waiting []*Lock
```

Comment:
Why not each node has a waiting queue?

Reply:
Waiting queue is moved from each node to the latch for those reasons:
1. nodes in the queue is inserted every now and then, if each node has is waiting queue, the queue would be created and destroyed. There will be more allocations, and it's less memory efficient.
2. I have an assumption that the waiting queue would not be large, when a list is small enough, an array is very efficient.
3. You may still remember the "first waiting one automatically become running" problem and the old code is complex enough to handle different states. If each node doesn't have the waiting queue, the problem could be avoid. 
@zhangjinpeng1987